pkg/gather/clusterconfig/version.go

```diff
@@ -2,75 +2,110 @@ package clusterconfig
 	if err != nil {
 		c <- gatherResult{nil, []error{err}}
 		return
 	}
-	c <- gatherResult{[]record.Record{{Name: "config/version", Item: record.JSONMarshaller{Object: anonymizeClusterVersion(config)}}}, nil}
-}
-
-func GetClusterVersion(ctx context.Context, kubeConfig *rest.Config) (*configv1.ClusterVersion, error) {
-	gatherConfigClient, err := configv1client.NewForConfig(kubeConfig)
+	gatherKubeClient, err := kubernetes.NewForConfig(g.gatherProtoKubeConfig)
 	if err != nil {
-		return nil, err
+		c <- gatherResult{nil, []error{err}}
+		return
 	}
-	config, err := gatherConfigClient.ClusterVersions().Get(ctx, "version", metav1.GetOptions{})
+	records, errors := getClusterVersion(g.ctx, gatherConfigClient, gatherKubeClient.CoreV1())
+	c <- gatherResult{records, errors}
+}
+
+func getClusterVersion(ctx context.Context, configClient configv1client.ConfigV1Interface, coreClient corev1client.CoreV1Interface) ([]record.Record, []error) {
+	config, err := configClient.ClusterVersions().Get(ctx, "version", metav1.GetOptions{})
 	if errors.IsNotFound(err) {
 		return nil, nil
 	}
 	if err != nil {
-		return nil, err
+		return nil, []error{err}
 	}
-	return config, nil
-}
 
-// GatherClusterID stores ClusterID from ClusterVersion version
-// This method uses data already collected by Get ClusterVersion. In particular field .Spec.ClusterID
-// The Kubernetes api https://github.com/openshift/client-go/blob/master/config/clientset/versioned/typed/config/v1/clusterversion.go#L50
-// Response see https://github.com/openshift/api/blob/master/config/v1/types_cluster_version.go#L38
-//
-// Location in archive: config/id/
-// See: docs/insights-archive-sample/config/id
-// Id in config: id
-func GatherClusterID(g *Gatherer, c chan<- gatherResult) {
-	defer close(c)
-	version, err := GetClusterVersion(g.ctx, g.gatherKubeConfig)
+	records := []record.Record{
+		{Name: "config/version", Item: record.JSONMarshaller{Object: anonymizeClusterVersion(config)}},
+	}
+
+	if config.Spec.ClusterID != "" {
+		records = append(records, record.Record{Name: "config/id", Item: marshal.Raw{Str: string(config.Spec.ClusterID)}})
+	}
+
+	// TODO: In the future, make this conditional on sad ClusterVersion conditions or ClusterVersionOperatorDown alerting, etc.
+	namespace := "openshift-cluster-version"
+	now := time.Now()
+	unhealthyPods := []*corev1.Pod{}
+
+	pods, err := coreClient.Pods(namespace).List(ctx, metav1.ListOptions{})
 	if err != nil {
-		c <- gatherResult{nil, []error{err}}
-		return
+		klog.V(2).Infof("Unable to find pods in namespace %s for cluster-version operator", namespace)
+		return records, nil
 	}
-	if version == nil {
-		c <- gatherResult{nil, nil}
-		return
+	for i := range pods.Items {
+		pod := &pods.Items[i]
+
+		// TODO: shift after IsHealthyPod
+		records = append(records, record.Record{Name: fmt.Sprintf("config/pod/%s/%s", pod.Namespace, pod.Name), Item: record.JSONMarshaller{Object: pod}})
```

Comment:
OK but what is the plan with the shift? It can always cause some problems if you remove some things that were already in the archive. I guess it's hopefully not a big deal in this case, but it's still good to know the plan :)

Reply:
Plan is:

1. Land this PR for [rhbz#1942271](https://bugzilla.redhat.com/show_bug.cgi?id=1942271).
2. Backport this PR to 4.7.
3. Collect a few weeks of Insights tarballs.
4. Audit them for CVO `tolerations`, to see if Insights-reporting users have been making any changes we're concerned about clobbering.
5. Follow-up insights PR to address the TODOs here, so we only collect the CVO pod when it's unhealthy.
6. In parallel with 5, patch the CVO to fix [rhbz#1941901](https://bugzilla.redhat.com/show_bug.cgi?id=1941901).

I'm agnostic about whether 5 gets a bug and a backport to 4.7.z.  Would be easy enough to hang on a third "insights operator collects CVO when the pod is healthy" bug series if we wanted to.

Reply:
OK thanks so there's plan to backport this to 4.7 as well if I understand it correctly right.